tests/Doctrine/Tests/ORM/Tools/Pagination/PaginatorTest.php

```diff
@@ -0,0 +1,70 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\Tests\ORM\Tools\Pagination;
+
+use Doctrine\ORM\Decorator\EntityManagerDecorator;
+use Doctrine\ORM\Internal\Hydration\AbstractHydrator;
+use Doctrine\ORM\Query;
+use Doctrine\ORM\Tools\Pagination\Paginator;
+use Doctrine\Tests\Mocks\ConnectionMock;
+use Doctrine\Tests\Mocks\DriverMock;
+use Doctrine\Tests\OrmTestCase;
+
+class PaginatorTest extends OrmTestCase
+{
+    public function testPaginationWithSubSelectStripsParametersInSelect()
+    {
+        $connection = $this->getMockBuilder(ConnectionMock::class)
+            ->setConstructorArgs([[], new DriverMock()])
+            ->setMethods(['executeQuery'])
+            ->getMock()
+        ;
+
+        $em = $this->getMockBuilder(EntityManagerDecorator::class)
+            ->setConstructorArgs([$this->_getTestEntityManager($connection)])
+            ->setMethods(['newHydrator'])
+            ->getMock()
+        ;
+
+        $abstractHydrator = $this->createMock(AbstractHydrator::class);
+        $em->method('newHydrator')->willReturn($abstractHydrator);
+
+        $paramInWhere     = 1;
+        $paramInSubSelect = 2;
+        $returnedIds      = [10];
+
+        $query = new Query($em);
+        $query->setDQL(
+            "SELECT u,
+                (
+                    SELECT MAX(a.version)
+                    FROM Doctrine\\Tests\\Models\\CMS\\CmsArticle a
+                    WHERE a.user = u AND 1 = :paramInSubSelect
+                ) AS HIDDEN max_version
+            FROM Doctrine\\Tests\\Models\\CMS\\CmsUser u
+            WHERE u.id = :paramInWhere"
+        );
+        $query->setParameters(compact('paramInWhere', 'paramInSubSelect'));
+        $paginator = (new Paginator($query, true))->setUseOutputWalkers(false);
+
+        $abstractHydrator->method('hydrateAll')->willReturn([$returnedIds]);
+        $connection->expects($this->exactly(3))->method('executeQuery');
+
+        for ($invocationIndex=0; $invocationIndex<2; $invocationIndex++) {
```

Comment:
Since this is repeated only 2 times, repeating the `expects()` is OK

Reply:
So what's the general rule when to duplicate code and when to put it to loop? 3 time repeat?

Reply:
It's a loop done for 2 iterations - seems easier to read with explicit `at(0)` `at(1)`, `at(2)`